### PR TITLE
playbook/role to update ISTs in the openshift namespace

### PR DESCRIPTION
**Pull request by gabemontero**
A very fine grained playbook/role.  And after some attempts, it seemed to orthogonal to this openshift-ansible newbie to somehow include in the openshift_examples role, but of course if reviewers advise otherwise I'll adjust.

This allows updates via `oc tag` for the current images used for any IST in the openshift namespace.

Possible uses:
1) As implied above, fine grained image update post initial install/deploy, but prior to any full scale upgrade of note ... or an ansible based path to `oc tag` in the openshift namespace
2) Facilitate api.ci/ci-operator based ci/cd flows for in house dev ... see item 4) in https://github.com/openshift/release/pull/947#issue-195288298

@bparees @smarterclayton @stevekuznetsov @sdodson ptal

### CODE REVIEW

File: roles/openshift_imagestreamtags/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+- name: Ensure openshift project exists
+  oc_project:
+    name: openshift
+    state: present
+    node_selector:
+    - ""
+
+- name: Tag image into an image stream in the openshift namespace
+  command: >
+    {{ openshift_client_binary }} tag --source=docker {{oreg_url}} --config={{ openshift.common.config_base }}/master/admin.kubeconfig openshift/{{ openshift_ist }}
```
Comment (bparees) [inline roles/openshift_imagestreamtags/tasks/main.yml]: oreg_url normally takes the form:
`registry.hostname.com/openshift3/ose-${component}:${version}`

(ie it's normally a template, not a specific image spec)

I don't see how you can use it in this way here?

In general if we're going to do this at all, i'd expect this to take two inputs:

1) an array of image sources
2) an array of ISTs to tag those sources as

File: playbooks/openshift-imagestreamtags/update_imagestreamtags.yml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+- import_playbook: ../init/main.yml
+  
+- name: Update image stream tag
+  hosts: all
+  roles:
+    - openshift_imagestreamtags
```
Comment (sdodson) [inline playbooks/openshift-imagestreamtags/update_imagestreamtags.yml]: We've tried to establish a pattern of only putting "entry point" playbooks in playbooks/foo/ and then putting actual implementation in playbooks/foo/private/. The expectation being that only other playbooks which have previously run init would call directly to /private/

So to do that here split lines 4-7 off into playbooks/openshift-imagestreams/private/update_imagestreamtags.yml and then add something like the following here.

```
- import_playbook: private/update_imagestreamtags.yml
```

Also, I think hosts should be `oo_first_master` which will ensure that this runs only on the first master. That's what we do with most things that just interact with the api.

Comment (sdodson) [reply playbooks/openshift-imagestreamtags/update_imagestreamtags.yml]: Maybe also this should live in the existing playbooks/openshift-master/ tree? We don't really have well defined patterns for how we've named top level playbook directories.

File: roles/openshift_imagestreamtags/tasks/main.yml (hunk 1 above)
Comment (gabemontero) [reply roles/openshift_imagestreamtags/tasks/main.yml]: We can certainly introduce a new environment variable if straying from the typical format of `oreg_url` is less desirable than introducing another variable.

@sdodson - any opinions there?

Comment (bparees) [reply roles/openshift_imagestreamtags/tasks/main.yml]: the bad part about reusing oreg_url but requiring a different format/value is it means someone has to edit their inventory file in between running playbooks.  That seems error prone.

**Comment by gabemontero** [comment]
On @bparees 's point regarding a change like this is `purely to support our CI system` .... certainly that scenario was the original impetus for my crafting of this.

But as I hinted at in the description, and as he and I discussed off line, while the need for such a capability may not be requested by customers yet, I do not think it is entirely far fetched that they would want an ansible enabled entryway into tagging in images.

Certainly in the context of the various "example" images in the openshift namespace, we've had occasion during customer support to have them import recent updates to images to either diagnose problems of incorporate fixes out of band from broader upgrades.  As such, they might start feeling more comfortable doing such tasks from an ansible based inventory of playbooks vs. manually logging into a cluster and invoking `oc`.

File: playbooks/openshift-imagestreamtags/update_imagestreamtags.yml (hunk 2 above)
Comment (gabemontero) [reply playbooks/openshift-imagestreamtags/update_imagestreamtags.yml]: OK, I'll push the first part in a bit.  On the where to live piece, yeah, I struggled on where to put this, ease of access, etc.

In the end, it didn't seem to fit with what I saw under openshift-master, but I fully admit I don't have enough experience / context in this space, and will ultimately defer if strong opinions emerge.

File: roles/openshift_imagestreamtags/tasks/main.yml (hunk 1 above)
Comment (gabemontero) [reply roles/openshift_imagestreamtags/tasks/main.yml]: ok i'm convinced ... will introduce new var

**Comment by gabemontero** [comment]
changes based on first round of comments pushed

**Comment by gabemontero** [comment]
[(8) Run Jenkins extended tests in Jenkins repo [tech_debt][jenkins_integration]](https://trello.com/c/l3kGmkza/1466-8-run-jenkins-extended-tests-in-jenkins-repo-techdebtjenkinsintegration)

File: roles/openshift_imagestreamtags/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+- name: Ensure openshift project exists
+  oc_project:
+    name: openshift
+    state: present
+    node_selector:
+    - ""
+
+- name: Tag image into an image stream in the openshift namespace
+  command: >
+    {{ openshift_client_binary }} tag --source=docker {{docker_image_url}} --config={{ openshift.common.config_base }}/master/admin.kubeconfig openshift/{{ openshift_ist }}
```
Comment (michaelgugino) [inline roles/openshift_imagestreamtags/tasks/main.yml]: docker_image_url is not a great name for a variable.

We should probably create a list/dict so users can pass in more than one without have to adjust their inventory between runs.
```yaml
openshift_image_stream_tags:
  some_ist: source_url
  some_other_ist: source2_url
```

Comment (gabemontero) [reply roles/openshift_imagestreamtags/tasks/main.yml]: update pushed - thanks @michaelgugino

**Comment by gabemontero** [comment]
ok ... @smarterclayton , @bparees , @stevekuznetsov , and I had pow-wow

Among many items we've ruled out the need for this as part of the trello card I referenced

Closing out.

**Pull request closed without merging**